static/panes/pane.ts

```diff
@@ -0,0 +1,197 @@
+
+    /**
+     * Emit analytics event for opening the pane tab. Typical implementation
+     * looks like this:
+     *
+     * ```ts
+     * ga.proxy('send', {
+     *   hitType: 'event',
+     *   eventCategory: 'OpenViewPane',
+     *   eventAction: 'RustMir',
+     * });
+     * ```
+     */
+    abstract registerOpeningAnalyticsEvent(): void
+
+    /** Optionally overridable code for initializing pane buttons */
+    registerButtons(state: BasePaneState /* typeof state */): void {}
+
+    /** Optionally overridable code for initializing event callbacks */
+    registerCallbacks(): void {}
+
+    abstract getPaneName(): string;
+    abstract onCompiler(id: number, compiler: any, options: any, editorId: number): void;
+    abstract onCompileResult(id: unknown, compiler: unknown, result: unknown): void;
+    abstract close(): void;
+
+    /** Initialize standard lifecycle hooks */
+    protected registerStandardCallbacks(): void {
+        this.fontScale.on('change', this.updateState, this);
+        this.container.on('destroy', this.close, this);
+        this.container.on('resize', this.resize, this);
+        this.eventHub.on('compileResult', this.onCompileResult, this);
+        this.eventHub.on('compiler', this.onCompiler, this);
+        this.eventHub.on('compilerClose', this.onCompilerClose, this);
+        this.eventHub.on('settingsChange', this.onSettingsChange, this);
+        this.eventHub.on('shown', this.resize, this);
+        this.eventHub.on('resize', this.resize, this);
+    }
+
+    protected setTitle() {
+        this.container.setTitle(this.getPaneName());
+    }
+
+    protected onCompilerClose(id: unknown) {
+        if (this.compilerInfo.compilerId === id) {
+            _.defer(() => this.container.close());
+        }
+    }
+
+    protected onDidChangeCursorSelection(event: monaco.editor.ICursorSelectionChangedEvent) {
+        if (this.isAwaitingInitialResults) {
+            this.selection = event.selection;
+            this.updateState();
+        }
+    }
+
+    protected onSettingsChange(settings: SiteSettings) {
+        this.settings = settings;
+        this.editor.updateOptions({
+            contextmenu: settings.useCustomContextMenu,
+            minimap: {
+                enabled: settings.showMinimap,
+            },
+            fontFamily: settings.editorsFFont,
+            fontLigatures: settings.editorsFLigatures,
+        });
+    }
+
+    getCurrentState() {
+        const state = {
+            id: this.compilerInfo.compilerId,
+            editorId: this.compilerInfo.editorId,
+            selection: this.selection,
+        };
+        this.fontScale.addState(state);
+        return state;
+    }
+
+    updateState() {
+        this.container.setState(this.getCurrentState());
+    }
+
+    resize() {
+        const topBarHeight = utils.updateAndCalcTopBarHeight(this.domRoot,
+            this.topBar, this.hideable);
+        this.editor.layout({
```

Comment:
Maybe check that editor is not null here, as there could be tools that don't use it